pre-bundling error when the defines contains expression

### Describe the bug

This is caused by the inconsistent behavior of vite and esbuild. How can we avoid this problem? @patak-js @evanw

From vite's [docs](https://vitejs.dev/config/#define)
> Starting from 2.0.0-beta.70, string values will be used as raw expressions, so if defining a string constant, it needs to be explicitly quoted (e.g. with JSON.stringify).

From esbuild's [docs](https://esbuild.github.io/api/#define)
> Replacement expressions must either be a JSON object (null, boolean, number, string, array, or object) or a single identifier.

```
> vite

Pre-bundling dependencies:
  vue/dist/vue.cjs
(this will be run only when your dependencies or config have changed)
 > error: Invalid define value (must be valid JSON syntax or a single identifier): 1 + 1

error when starting dev server:
Error: Build failed with 1 error:
error: Invalid define value (must be valid JSON syntax or a single identifier): 1 + 1
    at failureErrorWithLog (/Users/ygj6/Downloads/vite-qnxzcc/node_modules/esbuild/lib/main.js:1493:15)
    at /Users/ygj6/Downloads/vite-qnxzcc/node_modules/esbuild/lib/main.js:1151:28
    at runOnEndCallbacks (/Users/ygj6/Downloads/vite-qnxzcc/node_modules/esbuild/lib/main.js:941:63)
    at buildResponseToResult (/Users/ygj6/Downloads/vite-qnxzcc/node_modules/esbuild/lib/main.js:1149:7)
    at /Users/ygj6/Downloads/vite-qnxzcc/node_modules/esbuild/lib/main.js:1258:14
    at /Users/ygj6/Downloads/vite-qnxzcc/node_modules/esbuild/lib/main.js:629:9
    at handleIncomingPacket (/Users/ygj6/Downloads/vite-qnxzcc/node_modules/esbuild/lib/main.js:726:9)
    at Socket.readFromStdout (/Users/ygj6/Downloads/vite-qnxzcc/node_modules/esbuild/lib/main.js:596:7)
    at Socket.emit (events.js:376:20)
    at addChunk (internal/streams/readable.js:309:12)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! vite-project@0.0.0 dev: `vite`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the vite-project@0.0.0 dev script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/ygj6/.npm/_logs/2021-11-06T09_55_10_642Z-debug.log

Process finished with exit code 1
```

### Reproduction

https://stackblitz.com/edit/vite-qnxzcc

### System Info

```shell
System:
    OS: macOS 12.0.1
    CPU: (8) x64 Intel(R) Core(TM) i5-8279U CPU @ 2.40GHz
    Memory: 183.78 MB / 8.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.17.0 - /usr/local/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 6.14.13 - /usr/local/bin/npm
  Browsers:
    Chrome: 95.0.4638.69
    Safari: 15.1
  npmPackages:
    vite: ^2.6.0 => 2.6.13 
```


### Used Package Manager

yarn

### Logs

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to https://github.com/vuejs/vue-next instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/docs/config/index.md
+++ b/docs/config/index.md
@@ -156,6 +156,8 @@ export default defineConfig(({ command, mode }) => {
 
   - Starting from `2.0.0-beta.70`, string values will be used as raw expressions, so if defining a string constant, it needs to be explicitly quoted (e.g. with `JSON.stringify`).
 
+  - To be consistent with [esbuild behavior](https://esbuild.github.io/api/#define), expressions must either be a JSON object (null, boolean, number, string, array, or object) or a single identifier.
+
   - Replacements are performed only when the match is surrounded by word boundaries (`\b`).
 
   ::: warning

